### PR TITLE
Tracking user weekly cells-processed quotas

### PR DESCRIPTION
**Pull request by KevinCLydon**
Users now have an associated cell quota, which is a count of cells they can process during a week.  Their quota resets at midnight on the weekday that matches the day their account was created.

If an annotate request is received that would cause the user to go over their quota for the week, a 403 response is returned with a message that includes their remaining quota for the week.

To account for cells currently being processed, the user activity table now also tracks an event for each row, either STARTED, SUCCEEDED, or FAILED.  STARTED events are matched to SUCCEEDED/FAILED events by a request ID.  The rows are matched up when calculating the user's remaining quota.  Only SUCCEEDED and unmatched STARTED rows are counted.

There is also a new endpoint for getting a user's total and remaining quota, along with their quota reset date.

Some of my concerns with this PR:

- For unit tests, I mostly just modified existing annotate ones to make sure the quota is being checked and the correct info is being logged.  I'm undecided on whether this is sufficient.
- The query I'm using to check remaining quota is a bit ugly, but I'm not sure of a cleaner way to do it.  It seems to perform fine for now, but I'm a little concerned about how that performance will scale.
- I typically agonize over how best to organize my code to the point of paralysis, so I decided this time to just go with my gut.  As a result, I'm sure there's some stuff that could be cleaner or better organized.

### CODE REVIEW

File: src/casp/services/api/data_manager/cellarium_general.py [hunk 1]
```diff
@@ -114,7 +121,80 @@ def log_user_activity(self, user_id: int, model_name: str, method: str, cell_cou
...
+        if days_to_subtract < 0:
+            days_to_subtract += 7
+        start_of_week = datetime.datetime.today().replace(
+            hour=0, minute=0, second=0, microsecond=0
+        ) - datetime.timedelta(days=days_to_subtract)
+
+        with self.system_data_db_session_maker() as session:
+            cell_count_sum_query = sa.sql.text(
+                """
+                select sum(cells_processed) from(
```
Comment (nmalfroy) [inline src/casp/services/api/data_manager/cellarium_general.py:170]: I think this query will be more efficient and I think a little easier to read?
```
select
sum(cells_processed)
from (select distinct
          request_id,
          array_agg(event) events, -- aggregates all of the requests ids into an array
          max(cell_count) cells_processed -- presuming the start or end is 0 and the other is the cell count
      from users_useractivity ua
      where user_id = :id
        and finished_time >= :start_of_week
        and request_id is not null
      group by request_id
     ) requests
where not 'FAILED' = ANY(events);
```

Lemme know what you think!  I got the same results as with your query but with a lower optimizer cost

Comment (nmalfroy) [reply src/casp/services/api/data_manager/cellarium_general.py:170]: It could always benefit from indexes on finished and request_id if the perf gets rough too (I think it'll be totally fine for now)

Comment (nmalfroy) [reply src/casp/services/api/data_manager/cellarium_general.py:170]: I wrote my assumptions in the comments

File: src/casp/services/api/services/cell_operations_service.py [hunk 2]
```diff
@@ -58,6 +62,51 @@ def authorize_model_for_user(self, user: models.User, model_name: str) -> models
...
+
+        :param user: User object to check quota for.
+        :param file: Anndata file to check the number of cells in.
+        :param model_name: Model name to use for annotation.
+        :param method: Method name to log in the database.
+
+        :raises exceptions.QuotaExceededException: If the number of cells in the anndata file exceeds the user's quota.
+        """
+        cell_count = self.__get_cell_count_from_anndata(file=file)
+        with user_activity_lock:
```
Comment (nmalfroy) [inline src/casp/services/api/services/cell_operations_service.py:95]: Ah, good to know that's how to do locks!  I'll just mention that this is only per instance so if we're running several replicas, this won't actually lock

File: src/casp/services/api/services/cell_operations_service.py [hunk 3]
```diff
@@ -215,6 +264,13 @@ async def annotate_cell_type_summary_statistics_strategy(
 
         :return: JSON response with annotations.
         """
+        logger.info(
+            "Verifying quota and logging user activity for annotate_cell_type_summary_statistics_strategy start"
+        )
+        self.__verify_quota_and_log_activity(
+            user=user, file=file, model_name=model_name, method="annotate_cell_type_summary_statistics_strategy"
+        )
+        logger.info("Authorizing user")
```
[the remark this answers is not in the document]
Comment (KevinCLydon) [reply src/casp/services/api/services/cell_operations_service.py:273]: That's a good question.  Probably not?  I noticed it was in the other annotate method but missing from this one so I added it here while I was adding the verify quota stuff.  Although realistically I don't know that it's actually a super helpful thing to log

File: src/casp/services/api/services/exceptions.py [hunk 4]
```diff
@@ -25,3 +25,7 @@ class VectorSearchResponseError(APIInternalError):
 
 class CellMetadataColumnDoesNotExist(InvalidInputError):
     http_code: int = status.HTTP_400_BAD_REQUEST
+
+
+class QuotaExceededException(APIBaseException):
+    http_code: int = status.HTTP_403_FORBIDDEN
```
[the remark this answers is not in the document]
Comment (KevinCLydon) [reply src/casp/services/api/services/exceptions.py:31]: Oh that's a good point.  I'll change that

File: src/casp/services/api/routers/cell_operations_router.py [hunk 5]
```diff
@@ -33,13 +36,25 @@ async def annotate(
...
+    try:
+        return await cell_operations_service.annotate_cell_type_summary_statistics_strategy(
+            user=request_user,
+            file=file.file,
+            model_name=model_name,
+            include_extended_output=include_dev_metadata,
+        )
+    except Exception as e:
+        try:
+            cell_operations_service.cellarium_general_dm.log_user_activity(
```
[the remark this answers is not in the document]
Comment (KevinCLydon) [reply src/casp/services/api/routers/cell_operations_router.py:48]: Oh that's a good idea.  I'll see if I can put something like that together instead

File: src/casp/services/api/services/cell_operations_service.py (hunk 2 above)
Comment (KevinCLydon) [reply src/casp/services/api/services/cell_operations_service.py:95]: Oh I didn't think of that.  That does complicate things.  I'm not sure how to handle it in that case 🤔

**Review by fedorgrab — CHANGES_REQUESTED** [review]
A couple of small things: feel free to disregard my discussion about the locks. :)

File: src/casp/services/api/data_manager/cellarium_general.py [hunk 6]
```diff
@@ -114,7 +121,65 @@ def log_user_activity(self, user_id: int, model_name: str, method: str, cell_cou
...
+        if days_to_subtract < 0:
+            days_to_subtract += 7
+        start_of_week = datetime.datetime.today().replace(
+            hour=0, minute=0, second=0, microsecond=0
+        ) - datetime.timedelta(days=days_to_subtract)
+
+        with self.system_data_db_session_maker() as session:
+            cell_count_sum_query = sa.sql.text(
+                """
+                select
```
[the remark this answers is not in the document]
Comment (KevinCLydon) [reply src/casp/services/api/data_manager/cellarium_general.py:170]: That sounds reasonable.  I think I'll go the second route because I think the query is just complex enough that it would be a little hard to parse in SQLAlchemy form.

File: src/casp/services/api/services/cellarium_general_service.py [hunk 7]
```diff
@@ -69,3 +70,26 @@ def get_model_for_user(self, model_name: str, user: models.User) -> models.CASMo
...
+        Get user quota information
+
+        :param user: User object to check quota for
+
+        :return: User quota information
+        """
+
+        days_to_add = datetime.datetime.now().weekday() - user.created_at.weekday()
+        if days_to_add <= 0:
+            days_to_add += 7
```
[the remark this answers is not in the document]
Comment (KevinCLydon) [reply src/casp/services/api/services/cellarium_general_service.py:85]: I'm not really sure.  The logic right now for determining the start of the quota period would definitely break if we were to change that number.  Like, if we wanted to change the quota period to 2 weeks or 3 days or something, it would take more changes than just updating that number.

File: src/casp/services/api/services/cell_operations_service.py [hunk 8]
```diff
@@ -58,6 +59,52 @@ def authorize_model_for_user(self, user: models.User, model_name: str) -> models
...
 
+    def __get_cell_count_from_anndata(self, file: t.BinaryIO) -> int:
+        """
+        Get the number of cells in the anndata file.
+
+        :param file: Anndata file to get the number of cells from.
+
+        :return: Number of cells in the anndata file.
+        """
+        adata = anndata.read_h5ad(file)
```
[the remark this answers is not in the document]
Comment (KevinCLydon) [reply src/casp/services/api/services/cell_operations_service.py:70]: Oh neat, that's convenient.  I'll update that

[the remark this answers is not in the document]
Comment (KevinCLydon) [reply src/casp/services/api/services/cell_operations_service.py:70]: Reading in that mode doesn't seem to work with `BinaryIO`.  I assume we have to use `BinaryIO` because a fastapi `UploadFile` isn't the same thing as a python file object, right?

File: src/casp/services/api/data_manager/cellarium_general.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (fedorgrab) [reply src/casp/services/api/data_manager/cellarium_general.py]: Yep, that's a good point. Maybe marking somehow that it's a sqlalchemy text would also bea good idea? I believe this exact query couldn't be executed in pure SQL as it has SQLAlhemy specific string interpolation syntax. Smth like `file_name.sql.sa`? (Not sure about this idea myself, but it'd be great to hear your opinion).

File: src/casp/services/api/routers/cell_operations_router.py [hunk 9]
```diff
@@ -10,6 +11,8 @@
 cell_operations_service = services.CellOperationsService()
 cellarium_general_service = services.CellariumGeneralService()
 
+logger = logging.getLogger(__name__)
```
Comment (nmalfroy) [inline src/casp/services/api/routers/cell_operations_router.py:14]: I think this is maybe a straggler?

File: src/casp/services/api/services/cellarium_general_service.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (nmalfroy) [reply src/casp/services/api/services/cellarium_general_service.py:85]: That's a great point

**Review by nmalfroy — APPROVED** [review]
this looks good.  Great work!

File: src/casp/services/api/services/cell_operations_service.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (fedorgrab) [reply src/casp/services/api/services/cell_operations_service.py:70]: Looks like so, that's sad.:(

**Review by fedorgrab — APPROVED** [review]
Looks good to me.

File: src/casp/services/api/routers/cell_operations_router.py (hunk 9 above)
Comment (KevinCLydon) [reply src/casp/services/api/routers/cell_operations_router.py:14]: Oh yup, good catch
